### PR TITLE
feat: add PVForecastHomeAssistant provider

### LINKED ISSUE #1232: Support importing PV forecast time series from a Home Assistant entity via the HomeAssistant adapter
### Describe the enhancement or feature request:

EOS hat bereits einen HomeAssistant-Adapter (adapter.homeassistant) mit Feldern für Live-Messwerte (load_emr_entity_ids, grid_import_emr_entity_ids, pv_production_emr_entity_ids, device_measurement_entity_ids) und für das Zurückschreiben von Ergebnissen (solution_entity_ids, device_instruction_entity_ids). Es gibt aber kein Feld, um eine komplette PV-Prognose-Zeitreihe aus einer HA-Entity zu importieren.

Ich habe das sowohl über die REST-API (/openapi.json, Schema HomeAssistantAdapterCommonSettings-Input) als auch direkt im EOSdash-Config-UI geprüft: Unter pvforecast.provider stehen nur folgende Optionen zur Auswahl: PVForecastPVLib, PVForecastImport, PVForecastAkkudoktor, PVForecastVrm, PVNode, Solcast, ForecastSolar. Keine davon liest aus einer lokalen Home-Assistant-Entity – Solcast ist z. B. eine Cloud-API mit eigenem Key, kein HA-Bezug.

Problem: Home Assistant hat ein reifes Ökosystem an PV-Prognose-Integrationen (Solcast-Integration, Open-Meteo Solar Forecast, Helios-Forecast u. a.), die teils selbstlernend/kalibrierend sind und Topologien unterstützen, die EOS' eigener PVForecastPVLib-Provider aktuell nicht abbilden kann (z. B. mehrere Strings mit unterschiedlicher Ausrichtung an einem gemeinsamen Wechselrichter – siehe auch mein separates Bug-Issue zu PVForecastPVLib: fehlendes Clipping, ignorierte loss/userhorizon-Parameter, jede Plane bekommt einen eigenen vollständigen virtuellen Wechselrichter statt sich einen echten zu teilen). Nutzer, die EOS als Home-Assistant-Add-on betreiben, haben solche Prognosen oft schon vorliegen, können sie aber nicht direkt nutzen.

Aktueller Workaround (den ich gerade produktiv einsetze, funktioniert, ist aber unnötig umständlich):

pvforecast.provider auf PVForecastImport stellen
Eigene rest_command:-YAML-Definition in HA anlegen, die PUT /v1/prediction/import/PVForecastImport mit der Forecast-Zeitreihe der HA-Entity aufruft
Eigene Automation, die das alle 30 Min. auslöst
Das erfordert manuelle YAML-Bearbeitung (kein Config-Flow in HA verfügbar) und eine zusätzliche Automation außerhalb von EOS – für etwas, das der HA-Adapter eigentlich direkt könnte.

Vorschlag (bevorzugt, minimalinvasiv): Den HomeAssistant-Adapter so erweitern, dass er periodisch selbst eine konfigurierte Entity abfragt und intern das Äquivalent von PUT /v1/prediction/import/PVForecastImport ausführt – der Nutzer muss dann nur noch eine Entity-ID (bzw. ein Attribut davon) im Adapter eintragen, keine eigene rest_command + Automation mehr bauen.

Alternative (generischer, aufwändiger): Neues Feld im HomeAssistant-Adapter, z. B. pvforecast_entity_ids: list[str], das:

Vor jedem Prediction-Update die angegebene(n) HA-Entity(s) abfragt
Ein standardisiertes forecast-Attribut liest (Liste von {datetime, watts} o. ä. – ein Format, das viele HA-Wetter-/PV-Prognose-Integrationen bereits so oder ähnlich exponieren, z. B. das weather.get_forecasts-Format oder ein einfaches forecast: [{datetime, watts}, ...]-Attribut wie bei Solcast/Helios-Forecast)
Die Werte automatisch als pvforecast_ac_power in die Prediction-Pipeline einspeist (analog zu PVForecastImport, aber automatisch statt über eine externe Automation)
Das würde EOS' Wetter-/PV-Prognose-Layer als optional austauschbar gegen das reife HA-Ökosystem machen, ohne den eigenen PVForecastPVLib-Provider ersetzen zu müssen.

### Link to discussion and related issues

_No response_

### Proposed implementation

_No response_

### PR DESCRIPTION
**Pull request by matjhgc534z67umb**
## Summary

Adds a new `PVForecastHomeAssistant` PV forecast provider that reads a forecast
time series directly from a Home Assistant entity attribute — no external
`rest_command` + automation needed, unlike the current `PVForecastImport`-based
workaround.

Closes #1232.

> **Note:** This implementation (code, tests, and this PR description) was
> generated by an AI assistant (Claude), working from the discussion in #1232
> and directly against a live Home Assistant instance to verify the data
> contract. I've reviewed it and it passes the full existing test suite, but
> please review carefully as with any AI-assisted contribution.

## Why a new provider instead of extending the `HomeAssistant` adapter

`adapter.homeassistant` currently only syncs single scalar values (current
measurements, current instructions) — never a time series. PV forecasts are a
different kind of data: a list of future-dated points that need to become
`PVForecastDataRecord`s, which is exactly what `PVForecastProvider` already
models. `PVForecastHomeAssistant` follows the same self-polling pattern as
`PVForecastVrm` — its own settings, its own `_update_data()`, selected via
`pvforecast.provider` alongside the existing cloud providers — rather than
duplicating that machinery inside the adapter.

## Data contract

Reads a list of `{datetime, watts}` entries from a configurable entity
attribute (default: `forecast`). This matches the attribute shape already
exposed by Helios Forecast and is close to what Solcast exposes. Key names
and value unit (`W`/`kW`) are configurable (`datetime_key`, `value_key`,
`value_unit`) to accommodate integrations that use different field names.

Verified against a real Helios Forecast entity (`sensor.pv1_power_now`) in a
live installation — the response shape matches exactly:

```json
{"forecast": [{"datetime": "2026-08-18T06:30:00+02:00", "watts": 17.19}, ...]}
```

## Auth

- Running as a Home Assistant add-on: reuses the `SUPERVISOR_TOKEN` pattern
  already used by `adapter/homeassistant.py` — no extra config needed.
- Running standalone: `pvforecast.homeassistant.base_url` +
  `pvforecast.homeassistant.token` (long-lived access token).

## Test plan

- [x] New `tests/test_pvforecasthomeassistant.py` (7 tests): correct
      `pvforecast_ac_power` mapping, kW→W conversion, missing/empty
      attribute, malformed entries, HTTP error handling, missing-token error
- [x] `tests/test_prediction.py` provider-sequence/repr assertions updated
      for the new provider
- [x] `mypy` clean
- [x] `scripts/generate_config_md.py` runs cleanly, new field renders
      correctly in the generated config docs
- [x] Full existing test suite passes (no regressions)

### CODE REVIEW

File: src/akkudoktoreos/prediction/pvforecasthomeassistant.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+            ) from e
+        return response.json()
+
+    async def _update_data(self, force_update: Optional[bool] = False) -> None:
+        """Update forecast data in the PVForecastDataRecord format."""
+        settings = self.config.pvforecast.homeassistant
+        data = self._request_entity_state()
+        attributes = data.get("attributes", {})
+        forecast = attributes.get(settings.attribute)
+        if not forecast:
```
Comment (NormannK) [inline src/akkudoktoreos/prediction/pvforecasthomeassistant.py:129]: This early return leaves any future records stored by a previous successful refresh intact. The loop below also only upserts, so a shorter response or a malformed entry retains older values at omitted timestamps and EOS can optimize against a mixture of current and stale forecast data. Please parse/validate the response first and replace or clear the provider's active forecast window (or fail explicitly), and add a regression test that seeds old records before refreshing with an empty or shorter forecast.

File: src/akkudoktoreos/server/dash/uihints.py [hunk 2]
```diff
@@ -308,6 +308,10 @@ def register_lazy_options_source(key: str, source: LazyOptionsSource) -> None:
         item_path="pvforecast.planes",
         max_items_from="pvforecast.max_planes",
     ),
+    "pvforecast.homeassistant.entity_id": UiHint(
+        form="select",
+        options_from="adapter.homeassistant.homeassistant_entity_ids",
```
Comment (NormannK) [inline src/akkudoktoreos/server/dash/uihints.py]: This option source uses the existing Home Assistant adapter, whose entity lookup requires `SUPERVISOR_TOKEN`. In standalone mode, even when the new `base_url` and `token` settings are configured, the options resolve to an empty list and this select offers no manual-entry path. Please keep a text/hybrid input or populate the options through the provider's configured connection so standalone setup also works in EOSdash.

**Review by NormannK — CHANGES_REQUESTED** [review]
Thanks—this follows the architecture agreed in #1232, but the two issues below affect forecast correctness and the advertised standalone configuration. Requesting changes until they are addressed.

File: src/akkudoktoreos/prediction/pvforecasthomeassistant.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+                f"Entity '{settings.entity_id}' attribute '{settings.attribute}' contained no "
+                "usable forecast entries."
+            )
+            logger.error(error_msg)
+            raise ValueError(error_msg)
+
+        # Clear the whole active forecast window first, so a response that is shorter than a
+        # previous one (or has gaps) can't leave stale pvforecast_ac_power values behind at
+        # timestamps the new data no longer covers.
+        start_date = self.ems_start_datetime.start_of("day")
```
Comment (NormannK) [inline src/akkudoktoreos/prediction/pvforecasthomeassistant.py]: The active forecast begins at `ems_start_datetime`, but this clears from midnight. `PredictionProvider` deliberately retains records back to `keep_datetime` according to `prediction.historic_hours` (48 hours by default), so if Home Assistant exposes future-only rows, every refresh later in the day permanently blanks the retained forecast history between midnight and the EMS start. Please limit replacement to `start_date = self.ems_start_datetime` and use `self.end_datetime` for the end boundary so the existing DST adjustment is preserved. A regression test should seed a point before the EMS start and verify that it survives the refresh.

**Review by NormannK — COMMENTED** [review]
One additional issue in the stale-data fix:

File: src/akkudoktoreos/prediction/pvforecasthomeassistant.py (hunk 3 above)
Comment (NormannK) [reply src/akkudoktoreos/prediction/pvforecasthomeassistant.py]: Suggested implementation:

```python
start_date = self.ems_start_datetime
end_date = self.end_datetime

if start_date is None or end_date is None:
    raise RuntimeError("Cannot update PV forecast without a valid prediction window")

await self.key_delete_by_datetime(
    "pvforecast_ac_power",
    start_datetime=start_date,
    end_datetime=end_date,
)
```

This restricts replacement to the actual `[ems_start_datetime, end_datetime)` forecast horizon, preserves retained history, and reuses the existing DST-adjusted end boundary.

The regression test can seed both sides of the boundary:

```python
historical_dt = FIXED_EMS_START.subtract(hours=1)
stale_future_dt = FIXED_EMS_START.add(hours=5)

await pvforecast_instance.update_value(
    historical_dt, {"pvforecast_ac_power": 321.0}
)
await pvforecast_instance.update_value(
    stale_future_dt, {"pvforecast_ac_power": 999.0}
)

# Run _update_data() with a shorter forecast.

assert values[historical_dt] == 321.0
assert values[stale_future_dt] is None
```

That verifies historical data survives while omitted future values are still cleared.

**Review by NormannK — APPROVED** [review]

**Comment by NormannK** [comment]
very nice!
merged
